File: unipaith-backend/src/unipaith/schemas/prompt_library.py
"""Spec 42 §3.19–§3.20 / §4.17 — Prompt Library request/response schemas.

Shared by ``PromptLibraryService`` (which builds them) and
``api/prompt_library.py`` (which serves them). The frontend types in
``frontend/src/types/promptLibrary.ts`` are the wire mirror of these.
"""

from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from unipaith.models.prompt_library import (
    COMPETENCIES,
    CONFLICT_TYPES,
    DRAFT_STATUSES,
    ROLE_TYPES,
    STAKEHOLDER_TYPES,
)

# ── Catalog ──────────────────────────────────────────────────────────────────


class BehavioralPromptOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    prompt_key: str
    title: str
    intent_tag: str
    target_channel: str
    time_limit_seconds: int | None = None
    word_limit: int | None = None
    format_required: str
    evidence_required_flag: bool
    allowed_attachments_flag: bool
    language_option: str
    confidentiality_scope: str
    reuse_allowed_flag: str
    sort_order: int


# ── Responses (§3.19) ────────────────────────────────────────────────────────


class BehavioralResponseUpsert(BaseModel):
    response_text: str | None = Field(default=None, max_length=20000)
    draft_status: str = "draft"
    confidence_self_rating: int | None = Field(default=None, ge=1, le=5)
    needs_feedback_flag: bool = False
    linked_story_id: UUID | None = None

    def validated_draft(self) -> str:
        return self.draft_status if self.draft_status in DRAFT_STATUSES else "draft"


class StarFlags(BaseModel):
    situation: bool = False
    task: bool = False
    action: bool = False
    result: bool = False
    reflection: bool = False


class BehavioralResponseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    prompt_key: str
    response_text: str | None = None
    draft_status: str
    version_count: int
    last_edited: datetime | None = None
    confidence_self_rating: int | None = None
    authenticity_confidence_flag: bool
    needs_feedback_flag: bool
    reviewer_feedback_received_flag: bool
    star_situation_present: bool
    star_task_present: bool
    star_action_present: bool
    star_result_present: bool
    star_reflection_present: bool
    impact_metric_present: bool
    impact_metric_type: str | None = None
    impact_metric_value_band: str | None = None
    linked_story_id: UUID | None = None
    # SignalRecord metadata (§5).
    source: str
    confidence: int
    record_version: int
    updated_at: datetime


# ── Story bank (§3.20) ───────────────────────────────────────────────────────


class StoryBase(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    summary: str | None = Field(default=None, max_length=8000)
    primary_competency: str | None = None
    secondary_competency: str | None = None
    competency_tags: list[str] = Field(default_factory=list)
    context_tags: list[str] = Field(default_factory=list)
    role_type: str | None = None
    stakeholder_type: str | None = None
    conflict_type: str | None = None
    difficulty_tier: int | None = Field(default=None, ge=1, le=5)
    recency: date | None = None
    duration: str | None = Field(default=None, max_length=80)
    scale_tier: int | None = Field(default=None, ge=1, le=5)
    evidence_link: str | None = Field(default=None, max_length=500)
    referenceable_contact_flag: bool = False

    def _norm_enum(self, value: str | None, allowed: tuple[str, ...]) -> str | None:
        return value if value in allowed else None

    def cleaned(self) -> dict:
        """Drop out-of-vocabulary enum values to None (DB CHECK safety)."""
        data = self.model_dump()
        data["primary_competency"] = self._norm_enum(self.primary_competency, COMPETENCIES)
        data["secondary_competency"] = self._norm_enum(self.secondary_competency, COMPETENCIES)
        data["role_type"] = self._norm_enum(self.role_type, ROLE_TYPES)
        data["stakeholder_type"] = self._norm_enum(self.stakeholder_type, STAKEHOLDER_TYPES)
        data["conflict_type"] = self._norm_enum(self.conflict_type, CONFLICT_TYPES)
        data["competency_tags"] = [t for t in self.competency_tags if t in COMPETENCIES]
        return data


class StoryCreate(StoryBase):
    pass


class StoryUpdate(StoryBase):
    pass


class StoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    summary: str | None = None
    primary_competency: str | None = None
    secondary_competency: str | None = None
    competency_tags: list[str] = Field(default_factory=list)
    context_tags: list[str] = Field(default_factory=list)
    role_type: str | None = None
    stakeholder_type: str | None = None
    conflict_type: str | None = None
    difficulty_tier: int | None = None
    recency: date | None = None
    duration: str | None = None
    scale_tier: int | None = None
    evidence_link: str | None = None
    referenceable_contact_flag: bool
    source: str
    confidence: int
    record_version: int
    created_at: datetime
    updated_at: datetime


# ── Summary (§4.17 overlay + counts) ─────────────────────────────────────────


class PromptLibrarySummary(BaseModel):
    total_prompts: int
    answered_count: int
    final_count: int
    draft_count: int
    stories_count: int
    # §4.17 inference overlay — populated only when ai_prompt_library_v2_enabled.
    inference_enabled: bool
    interview_readiness_band: str | None = None
    interview_readiness_score: int | None = None
    readiness_detail: dict | None = None
    competency_coverage_map: dict | None = None
    competency_coverage_gaps: list[str] | None = None
    story_prompt_matching_table: list[dict] | None = None
    revision_priority_list: list[dict] | None = None
    suggested_practice_plan: str | None = None

File: frontend/src/types/promptLibrary.ts
// Spec 42 §3.19–§3.20 / §4.17 — Prompt Library + Story Bank wire types.
// Mirror of unipaith-backend/src/unipaith/schemas/prompt_library.py.

export type TargetChannel = 'interview' | 'essay' | 'short_answer' | 'video'
export type FormatRequired = 'STAR' | 'CAR' | 'freeform'
export type DraftStatus = 'none' | 'draft' | 'revised' | 'final'
export type ReadinessBand = 'low' | 'medium' | 'high'

export interface BehavioralPrompt {
  prompt_key: string
  title: string
  intent_tag: string
  target_channel: TargetChannel
  time_limit_seconds: number | null
  word_limit: number | null
  format_required: FormatRequired
  evidence_required_flag: boolean
  allowed_attachments_flag: boolean
  language_option: string
  confidentiality_scope: string
  reuse_allowed_flag: 'core' | 'school_specific'
  sort_order: number
}

export interface BehavioralResponse {
  prompt_key: string
  response_text: string | null
  draft_status: DraftStatus
  version_count: number
  last_edited: string | null
  confidence_self_rating: number | null
  authenticity_confidence_flag: boolean
  needs_feedback_flag: boolean
  reviewer_feedback_received_flag: boolean
  star_situation_present: boolean
  star_task_present: boolean
  star_action_present: boolean
  star_result_present: boolean
  star_reflection_present: boolean
  impact_metric_present: boolean
  impact_metric_type: string | null
  impact_metric_value_band: string | null
  linked_story_id: string | null
  source: string
  confidence: number
  record_version: number
  updated_at: string
}

export interface BehavioralResponseUpsert {
  response_text?: string | null
  draft_status?: DraftStatus
  confidence_self_rating?: number | null
  needs_feedback_flag?: boolean
  linked_story_id?: string | null
}

export interface Story {
  id: string
  title: string
  summary: string | null
  primary_competency: string | null
  secondary_competency: string | null
  competency_tags: string[]
  context_tags: string[]
  role_type: string | null
  stakeholder_type: string | null
  conflict_type: string | null
  difficulty_tier: number | null
  recency: string | null
  duration: string | null
  scale_tier: number | null
  evidence_link: string | null
  referenceable_contact_flag: boolean
  source: string
  confidence: number
  record_version: number
  created_at: string
  updated_at: string
}

export interface StoryInput {
  title: string
  summary?: string | null
  primary_competency?: string | null
  secondary_competency?: string | null
  competency_tags?: string[]
  context_tags?: string[]
  role_type?: string | null
  stakeholder_type?: string | null
  conflict_type?: string | null
  difficulty_tier?: number | null
  duration?: string | null
  scale_tier?: number | null
  evidence_link?: string | null
  referenceable_contact_flag?: boolean
}

export interface StoryMatch {
  prompt_key: string
  prompt_title: string
  best_story_id: string
  best_story_title: string
  score: number
}

export interface RevisionItem {
  prompt_key: string
  prompt_title: string
  strength: number
  reason: string
}

export interface PromptLibrarySummary {
  total_prompts: number
  answered_count: number
  final_count: number
  draft_count: number
  stories_count: number
  inference_enabled: boolean
  interview_readiness_band: ReadinessBand | null
  interview_readiness_score: number | null
  readiness_detail: {
    band: string
    score: number
    answered: number
    core_total: number
    star_avg: number
  } | null
  competency_coverage_map: Record<string, number> | null
  competency_coverage_gaps: string[] | null
  story_prompt_matching_table: StoryMatch[] | null
  revision_priority_list: RevisionItem[] | null
  suggested_practice_plan: string | null
}
